pkg/controllers/externaloidc/externaloidc_controller.go

```diff
@@ -0,0 +1,380 @@
+			} else if seenAudiences.Has(aud) {
+				errs = append(errs, fmt.Errorf("duplicate audience: %s", aud))
+			}
+
+			seenAudiences.Insert(aud)
+		}
+
+		var caCertPool *x509.CertPool
+		caCertPoolValid := true
+		if len(jwt.Issuer.CertificateAuthority) > 0 {
+			var err error
+			caCertPool, err = cert.NewPoolFromBytes([]byte(jwt.Issuer.CertificateAuthority))
+			if err != nil {
+				caCertPoolValid = false
+				errs = append(errs, fmt.Errorf("issuer CA is invalid: %v", err))
+			}
+		}
+
+		if issuerURL != nil && issuerURLValid && caCertPoolValid {
+			// make sure we can access the issuer with the given cert pool (system CAs used if pool is empty)
+			if err := validateCACert(*issuerURL, caCertPool); err != nil {
+				certMessage := "using the specified CA cert"
+				if caCertPool == nil {
+					certMessage = "using the system CAs"
+				}
+				errs = append(errs, fmt.Errorf("could not validate IDP URL %s: %v", certMessage, err))
+			}
+		}
+
+		seenClaims := sets.NewString()
+		for i, rule := range jwt.ClaimValidationRules {
+			if len(rule.Claim) == 0 {
+				errs = append(errs, fmt.Errorf("claim must not be empty for claim validation rule at index %d", i))
+			} else if seenClaims.Has(rule.Claim) {
+				errs = append(errs, fmt.Errorf("duplicate claim validation rule: %s", rule.Claim))
+			}
+
+			seenClaims.Insert(rule.Claim)
+		}
+
+		if len(jwt.ClaimMappings.Username.Claim) == 0 {
+			errs = append(errs, fmt.Errorf("username claim must not be empty"))
+		} else if jwt.ClaimMappings.Username.Prefix == nil {
+			errs = append(errs, fmt.Errorf("username prefix must not be nil when claim is set"))
+		}
+
+		if len(jwt.ClaimMappings.Groups.Claim) > 0 && jwt.ClaimMappings.Groups.Prefix == nil {
+			errs = append(errs, fmt.Errorf("group prefix must not be nil when claim is set"))
+		}
+	}
+
+	return
+}
+
+// validateCACert makes a request to the provider's well-known endpoint using the
+// specified CA cert pool to validate that the certs in the pool match the host
+func validateCACert(hostURL url.URL, caCertPool *x509.CertPool) error {
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{RootCAs: caCertPool},
+		},
+		Timeout: 5 * time.Second,
+	}
+
+	wellKnown := strings.TrimSuffix(hostURL.String(), "/") + oidcDiscoveryEndpointPath
+	req, err := http.NewRequest(http.MethodGet, wellKnown, nil)
+	if err != nil {
+		return fmt.Errorf("could not create well-known HTTP request: %v", err)
+	}
+
+	var resp *http.Response
+	var connErr error
+	retryCtx, cancel := context.WithTimeout(req.Context(), 10*time.Second)
+	retry.RetryOnConnectionErrors(retryCtx, func(ctx context.Context) (done bool, err error) {
+		resp, connErr = client.Do(req)
+		return connErr == nil, connErr
+	})
+	cancel()
+	if connErr != nil {
+		return fmt.Errorf("GET well-known error: %v", connErr)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
```

Comment:
Technically we only need to read the body, if the `resp.StatusCode` is NOT `http.StatusOK`, so I would only do it in the err case.

I never used it, but we might be even fine with http.MethodHEAD? Not sure, we could just use get and read the Body on err.

Reply:
>Technically we only need to read the body, if the `resp.StatusCode` is NOT `http.StatusOK`, so I would only do it in the err case.

I wrote it like this because depending on the status code, sometimes the body can contain useful information about the respective code.

>I never used it, but we might be even fine with http.MethodHEAD? Not sure, we could just use get and read the Body on err.

I chose GET because:
- this is the one defined in the discovery endpoint spec: https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationRequest
- this is the one that KAS uses to do the same validation, so we're preserving the behavior

HEAD is risky in the sense that it might be blocked on some servers as it's not defined in the spec.

Reply:
Good point, fixed; reading body only if we need to put it in an error message.